kite-data/kite-data-core/src/main/java/org/kitesdk/data/DatasetRecordException.java

```diff
@@ -0,0 +1,36 @@
+/*
+ * Copyright 2013 Cloudera Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.kitesdk.data;
+
+/**
+ * A DatasetException that signals that an individual record has failed.
+ * <p>
+ * When writing, this exception indicates that a record could not be written,
+ * but the writer is still valid and ready to write other records.
+ * <p>
+ * When reading, this exception indicates that a record could not be built from
+ * record data, but that reader is still valid and able to read other records.
```

Comment:
This implies, if thrown, that the reader has skipped the bad record. This is for formats like CSV and JSON, where the data may not have a valid record. CSV, for example, might be missing a required field. In that case, we can still read the next record because it might have valid data.

JSON is a similar case, where we could be missing a field or even have the wrong root type, but the iterator provides JsonNode instances, so we can assume that the next one will work.